R/separate-wider.R

```diff
@@ -0,0 +1,598 @@
+        (too_few == "debug" & width < expected_width) |
+        (too_many == "debug" & width > expected_width)
+      )
+
+    out[[debug_name(col, names_sep, "width")]] <- width
+    out[[debug_name(col, names_sep, "remainder")]] <- stringr::str_sub(x, expected_width + 1, width)
+    out[[debug_name(col, names_sep, "ok")]] <- !problem
+  }
+
+
+  out
+}
+
+#' @rdname separate_wider_delim
+#' @param patterns A named character vector where the names become column names
+#'   and the values are regular expressions that match the contents of the
+#'   vector. Unnamed components will match, but not be included in the output.
+#' @export
+separate_wider_regex <- function(
+    data,
+    cols,
+    patterns,
+    ...,
+    names_sep = NULL,
+    names_repair = "check_unique",
+    too_few = c("error", "debug", "start"),
+    col_remove = TRUE
+) {
+  check_installed("stringr")
+  check_required(cols)
+  check_character(patterns)
+  if (all(!have_name(patterns))) {
+    cli::cli_abort("{.arg patterns} must be a named character vector.")
+  }
+  check_dots_empty()
+  check_string(names_sep, allow_null = TRUE)
+  too_few <- arg_match(too_few)
+  check_bool(col_remove)
+
+  error_call <- current_env()
+
+  map_unpack(
+    data, {{ cols }},
+    function(x, col) str_separate_wider_regex(x, col,
+      patterns = patterns,
+      names_sep = names_sep,
+      too_few = too_few,
+      col_remove = col_remove,
+      error_call = error_call
+    ),
+    names_sep = names_sep,
+    names_repair = names_repair
+  )
+}
+
+str_separate_wider_regex <- function(x,
+                                     col,
+                                     patterns,
+                                     names_sep = NULL,
+                                     too_few = "error",
+                                     col_remove = TRUE,
+                                     error_call = caller_env()) {
+  has_name <- names2(patterns) != ""
+  groups <- paste0("(", ifelse(has_name, "", "?:"), patterns, ")")
+  full_match <- paste0("^", paste(groups, collapse = ""), "$")
+
+  match <- stringr::str_match(x, full_match)
+  if (ncol(match) != sum(has_name) + 1L) {
+    cli::cli_abort(c(
+      "Invalid number of groups.",
+      i = "Did you use () instead of (?:) inside {.arg patterns}?"
+    ), call = error_call)
+  }
+
+  matches <- match[, -1, drop = FALSE]
+  out <- as_tibble(matches, .name_repair = "none")
+  colnames(out) <- names2(patterns)[has_name]
+
+  if (!col_remove || too_few == "debug") {
+    out[[col]] <- x
+  }
+
+  match_count <- ifelse(is.na(x), NA, length(patterns))
+  remainder <- ifelse(is.na(x), NA, "")
+  problems <- !is.na(x) & is.na(match[, 1])
+  no_match <- which(problems)
+
+  if (length(no_match) > 0) {
+
+    if (too_few == "error") {
+      cli::cli_abort(c(
+        "Expected each value of {.var {col}} to match the pattern, the whole pattern, and nothing but the pattern.",
+        "!" = "{length(no_match)} value{?s} {?has/had} problem{?s}.",
```

Comment:
```suggestion
        "!" = "{length(no_match)} value{?s} {?has/have} problem{?s}.",
```

?, i.e. `5 values have problems` in the present tense, like `has`, not past tense `had`